### PR TITLE
Register Trees

### PR DESCRIPTION
**Pull request by sarveshpro**
Add register trees to user profile

### CODE REVIEW

**Comment by sagararyal** [comment]
@sarveshpro there is a bug that does not allow more than 50 trees. 

if I type 100, it is replaced with 50

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+  React.useEffect(() => {
+    if (localStorage.getItem('language')) {
+      let userLang = localStorage.getItem('language');
+      if (userLang) setUserLang(userLang);
+    }
+
+    async function getUserCountryBbox() {
+      var country = getStoredConfig('country');
+      const result = await fetch(`http://api.mapbox.com/geocoding/v5/mapbox.places/${country}.json?types=country&limit=1&access_token=${process.env.MAPBOXGL_ACCESS_TOKEN}`);
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:69]: Not using HTTPS here fails when using the website with SSL: `RegisterTrees.tsx?ed5c:69 Mixed Content: The page at 'https://localhost/register-trees' was loaded over HTTPS, but requested an insecure resource 'http://api.mapbox.com/geocoding/v5/mapbox.places/DE.json?types=country&limit=1&access_token=...'. This request has been blocked; the content must be served over HTTPS.`

Comment (norbertschuler) [reply src/features/user/UserProfile/components/RegisterTrees.tsx:69]: We never before used a direct API call for Mapbox - isn't this functionality available by any npm module we already use?

**Review by norbertschuler — CHANGES_REQUESTED** [review]

**Comment by norbertschuler** [comment]
- [ ] If I try to register without changes `0` trees, the focus is set again to the input field, but no error message is shown in the UI - maybe set initial value to `1` and do not allow `0` at all?

**Comment by norbertschuler** [comment]
- [ ] Uploading big photos can spill the UI that much, that you cannot hit the close icon any more. Maybe scale the images to a common size (keep aspect ration of course)?

Bildschirmfoto 2020-11-30 um 10 58 57

File: pages/c/[id].tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                        : `https://cdn.plant-for-the-planet.org/media/images/app/bg_layer.jpg`
+                }
+            >
+                {currentUserSlug ?
+                    <SingleContribution {...ContributionProps} />
+                    : null}
+            </LandingSection>
+
+        </>
+    ) : (<h2>NO Project ID FOUND</h2>) :
```
Comment (norbertschuler) [inline pages/c/[id].tsx:91]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                setErrorMessage(res.message);
+                setRegistered(false);
+              }
+            }
+          }
+        );
+
+        // handleNext();
+      } else {
+        setErrorMessage('select location on map');
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:160]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+      }
+    } else {
+      setErrorMessage("Something went wrong. Please contact support@plant-for-the-planet.org");
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:163]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+              <b> {t('me:registerTrees')} </b>
+            </h2>
+            <form onSubmit={handleSubmit(submitRegisterTrees)}>
+              <div className={styles.formField}>
+                <div className={styles.formFieldHalf}>
+                  <MaterialTextField
+                    inputRef={register({
+                      required: {
+                        value: true,
+                        message: 'Number of Trees is required',
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:201]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+                      },
+                      validate: (value) => parseInt(value, 10) >= 1,
+                    })}
+                    onInput={(e) => {
+                      e.target.value = e.target.value.replace(/[^0-9]/g, '');
+                    }}
+                    onChange={onTreeCountChange}
+                    label="Number of Trees"
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:209]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+                    />
+                  </MuiPickersUtilsProvider>
+                </div>
+              </div>
+              <div className={styles.formFieldLarge}>
+                <MaterialTextField
+                  inputRef={register({
+                    required: {
+                      value: true,
+                      message: 'Species is required',
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:248]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+                  name="species"
+                />
+                {errors.species && (
+                  <span className={styles.formErrors}>{errors.species.message}</span>
+                )}
+              </div>
+              <div className={styles.mapNote}>
+                {isMultiple ?
+                  <p>Draw a polygon on the map</p>
+                  : <p>Click on the map to mark a location</p>}
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:262]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+              </div>
+
+              {/* {errorMessage !== '' ? */}
+              <div className={`${styles.formFieldLarge} ${styles.center}`}><p className={styles.formErrors}>{`${errorMessage}`}</p></div>
+              {/* : null
+              } */}
+              <div className={styles.nextButton}>
+                <div onClick={handleSubmit(submitRegisterTrees)} className={styles.continueButton}>  {isUploadingData ? (
+                  <div className={styles.spinner}></div>
+                ) : 'Register'}</div>
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:318]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                            setIsUploadingData(false);
+                            setErrorMessage(res.message);
+                        }
+                    }
+                }
+            );
+
+            // handleNext();
+        } else {
+            setErrorMessage('select location on map');
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx:87]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+    return (
+        <form onSubmit={handleSubmit(submitRegisterTrees)}>
+            <div className={styles.formField}>
+                <div className={styles.formFieldHalf}>
+                    <MaterialTextField
+                        inputRef={register({
+                            required: {
+                                value: true,
+                                message: 'Number of Trees is required',
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx:99]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+                            },
+                            validate: (value) => parseInt(value, 10) >= 1 && parseInt(value, 10) <= 50,
+                        })}
+                        onInput={(e) => {
+                            e.target.value = e.target.value.replace(/[^0-9]/g, '');
+                            e.target.value = e.target.value > 50 ? 50 : e.target.value;
+                        }}
+                        onChange={onTreeCountChange}
+                        label="Number of Trees"
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx:108]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+                        <span className={styles.formErrors}>{errors.treeCount.message}</span>
+                    )}
+                </div>
+                <div className={styles.formFieldHalf}>
+                    <MuiPickersUtilsProvider utils={DateFnsUtils} locale={localeMapForDate[lang] ? localeMapForDate[lang] : localeMapForDate['en']}>
+                        <Controller
+                            render={props => (
+
+                                <DatePicker
+                                    label="Date Planted"
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx:123]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+                        />
+                    </MuiPickersUtilsProvider>
+                </div>
+            </div>
+            <div className={styles.formFieldLarge}>
+                <MaterialTextField
+                    inputRef={register({
+                        required: {
+                            value: true,
+                            message: 'Species is required',
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx:147]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+                        },
+                    })}
+                    label="Tree Species"
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx:150]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+                    variant="outlined"
+                    name="species"
+                />
+                {errors.species && (
+                    <span className={styles.formErrors}>{errors.species.message}</span>
+                )}
+            </div>
+            {isMultiple ?
+                <p>Draw a polygon on the map</p>
+                : <p>Click on the map to mark a location</p>}
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/BasicDetails.tsx:160]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+                    <span className={styles.formErrors}>{errors.treeCount.message}</span>
+                  )}
+                </div>
+                <div className={styles.formFieldHalf}>
+                  <MuiPickersUtilsProvider utils={DateFnsUtils} locale={localeMapForDate[userLang] ? localeMapForDate[userLang] : localeMapForDate['en']}>
+                    <Controller
+                      render={props => (
+
+                        <DatePicker
+                          label="Date Planted"
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:224]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees.tsx [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+              </div>
+              <div className={styles.formFieldLarge}>
+                <MaterialTextField
+                  inputRef={register({
+                    required: {
+                      value: true,
+                      message: 'Species is required',
+                    },
+                  })}
+                  label="Tree Species"
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees.tsx:251]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/SingleContribution.tsx [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+                className={styles.closeButton}
+                onClick={() => {
+                    router.push(`/t/${currentUserSlug}`, undefined, { shallow: true });
+                }}
+            >
+                <CancelIcon />
+            </div>
+            <h2 className={styles.contribTitle}>
+
+                <b> You planted {contribution.treeCount ? contribution.treeCount : '{null}'} {contribution.treeCount === 1 ? 'tree' : 'trees'} on {formatDate(contribution.plantDate)} </b>
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/SingleContribution.tsx:45]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/SingleContribution.tsx [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+            </h2>
+            <div className={styles.contributionInfo}>
+
+                <div className={styles.infoRow}>
+                    <div className={styles.infoItem}>
+                        <div className={styles.contributionTitle}>Contribution Information</div>
+                        <div className={styles.infoLine}><b>Species : </b>{contribution.treeSpecies}</div>
+                        <div className={styles.infoLine}><b>No of Trees : </b>{contribution.treeCount}</div>
+                        <div className={styles.infoLine}><b>Date : </b>{contribution.plantDate}</div>
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/SingleContribution.tsx:54]: i18n missing

Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/SingleContribution.tsx:54]: Please localize this date as well:

Bildschirmfoto 2020-11-30 um 11 07 07

File: src/features/user/UserProfile/components/RegisterTrees/UploadImages.tsx [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+            if (!res.code) {
+                let newUploadedImages = uploadedImages;
+                newUploadedImages.push(res)
+                setUploadedImages(newUploadedImages)
+                setIsUploadingData(false)
+                setErrorMessage(null)
+            } else {
+                if (res.code === 404) {
+                    setIsUploadingData(false)
+                    setErrorMessage('Contribution not found')
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/UploadImages.tsx:54]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/UploadImages.tsx [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+                }
+                else {
+                    setIsUploadingData(false)
+                    setErrorMessage('Error Occured')
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/UploadImages.tsx:58]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/UploadImages.tsx [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+                </div>
+                : null}
+            <div className={styles.formFieldLarge}>
+                <label htmlFor="upload" className={styles.fileUploadContainer} {...getRootProps()}>
+                    <AnimatedButton
+                        onClick={uploadPhotos}
+                        className={styles.continueButton}
+                    >
+                        <input {...getInputProps()} />
+                        {isUploadingData ? <div className={styles.spinner}></div> : 'Upload Photos'}
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/UploadImages.tsx:114]: i18n missing

File: src/features/user/UserProfile/components/RegisterTrees/UploadImages.tsx [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+
+                    </AnimatedButton>
+                    <p style={{ marginTop: '18px' }}>
+                        or drag here
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/RegisterTrees/UploadImages.tsx:118]: i18n missing

**Review by norbertschuler — CHANGES_REQUESTED** [review]
Please localize the code.

File: src/utils/getConfig/getStoredConfig.ts [hunk 25]
```diff
@@ -0,0 +1,39 @@
+import tenantConfig from '../../../tenant.config';
```
Comment (norbertschuler) [inline src/utils/getConfig/getStoredConfig.ts:1]: This code and the file https://github.com/Plant-for-the-Planet-org/planet-webapp/blob/develop/src/utils/storeConfig.ts should be together in one file and not separated. Also we should think about if we really want to store the whole config, but also the single values for country as `countryCode` and currency as `currencyCode` as own values in localstorage.

**Review by norbertschuler — DISMISSED** [review]
Config handling is not consistent.

File: src/utils/getConfig/getStoredConfig.ts (hunk 25 above)
Comment (sarveshpro) [reply src/utils/getConfig/getStoredConfig.ts]: first part done and created a new issue #515 to discuss the second part.

File: src/features/user/UserProfile/components/RegisterTrees.tsx (hunk 1 above)
Comment (sarveshpro) [reply src/features/user/UserProfile/components/RegisterTrees.tsx]: do we have to support users using HTTP?
no, the npm modules only have map libraries. there's a different library for geocoding API `@mapbox/mapbox-gl-geocoder`

Comment (norbertschuler) [reply src/features/user/UserProfile/components/RegisterTrees.tsx]: HTTP would only be useful for development running with `npm run dev` - I am using HTTPS locally by using a local proxy server called caddy to work with SSL.

**Comment by norbertschuler** [comment]
> * [x]  If I try to register without changes `0` trees, the focus is set again to the input field, but no error message is shown in the UI - maybe set initial value to `1` and do not allow `0` at all?

@sarveshpro Can't you use a number field like we use it for entering the target number at https://github.com/Plant-for-the-Planet-org/planet-webapp/blob/7626da74121fdf8fd6b0998e574ba49fcdf40220/src/features/user/UserProfile/components/AddTargetModal.tsx#L66-L81?

**Comment by norbertschuler** [comment]
- [ ] Uploading an image, deleting it, uploading it again duplicates the image (showing also the formerly deleted one) - I even did this twice resulting into showing three images at the end - if you try to delete an image already deleted at the backend, the API will return an error:

Bildschirmfoto 2020-11-30 um 16 12 27

**Comment by sarveshpro** [comment]
> > * [ ]   If I try to register without changes `0` trees, the focus is set again to the input field, but no error message is shown in the UI - maybe set initial value to `1` and do not allow `0` at all?
> 
> @sarveshpro Can't you use a number field like we use it for entering the target number at
> 
> https://github.com/Plant-for-the-Planet-org/planet-webapp/blob/7626da74121fdf8fd6b0998e574ba49fcdf40220/src/features/user/UserProfile/components/AddTargetModal.tsx#L66-L81
> 
> ?

I can but the value can change if the user scrolls on the field.

**Comment by norbertschuler** [comment]
> I can but the value can change if the user scrolls on the field.

I don't mind, but can you then add an error message if the user sends the form with `0` trees and is focussed again in this input field to make clear that he had done an input error?

Bildschirmfoto 2020-11-30 um 16 38 16

**Review by norbertschuler — APPROVED** [review]
LGFM

**Pull request merged**